src/swupd/fullfiles.go

```diff
@@ -0,0 +1,276 @@
+	if err != nil {
+		return fmt.Errorf("couldn't archive the file %s: %s", input, err)
+	}
+	uncompressedSize, err := uncompressed.Seek(0, io.SeekEnd)
+	if err != nil {
+		return fmt.Errorf("couldn't find the size of uncompressed fullfile: %s", err)
+	}
+	err = uncompressed.Sync()
+	if err != nil {
+		return fmt.Errorf("couldn't sync uncompressed fullfile: %s", err)
+	}
+	defer uncompressed.Close()
+
+	if DebugFullfiles {
+		log.Printf("DEBUG: Creating fullfile %s for regular file %s (%d bytes)", name, input, fi.Size())
+		log.Printf("DEBUG: %s (%d bytes, uncompressed)", filepath.Base(output), uncompressedSize)
+	}
+
+	// Pick the best compression option (or no compression) for that specific fullfile.
+	best := ""
+	var bestSize int64 = uncompressedSize
+	for i, c := range fullfileCompressors {
+		candidate := fmt.Sprintf("%s.%d.%s", output, i, c.Name)
+		_, err := uncompressed.Seek(0, io.SeekStart)
+		if err != nil {
+			return fmt.Errorf("couldn't seek in fullfile %s: %s", input, err)
+		}
+		args := fullfileCompressArgs{
+			output:       candidate,
+			tarredReader: uncompressed,
+		}
+		size, err := c.Func(args)
+		if err != nil {
+			log.Printf("WARNING: couldn't compress %s using compressor %q: %s", input, c.Name, err)
+			os.RemoveAll(candidate)
+			continue
+		}
+		if size < bestSize {
+			if best != "" {
+				os.RemoveAll(best)
+			}
+			best = candidate
+			bestSize = size
+		} else {
+			os.RemoveAll(candidate)
+		}
+
+		if DebugFullfiles {
+			log.Printf("DEBUG: %s (%d bytes)", filepath.Base(candidate), size)
+		}
+	}
+
+	if best != "" {
+		// Failure during rename might indicate some further problems, so return error
+		// instead of ignoring it (since there is an uncompressed version).
+		return os.Rename(best, output)
+	}
+	return nil
+}
+
+func tarRegularFullfile(w io.Writer, input, name string, fi os.FileInfo) error {
+	tw := tar.NewWriter(w)
+	hdr, err := getHeaderFromFileInfo(fi)
+	if err != nil {
+		return err
+	}
+	hdr.Typeflag = tar.TypeReg
+	hdr.Name = name
+	err = tw.WriteHeader(hdr)
+	if err != nil {
+		return err
+	}
+	in, err := os.Open(input)
+	if err != nil {
+		return err
+	}
+	_, err = io.Copy(tw, in)
+	if err != nil {
+		in.Close()
+		return err
+	}
+	in.Close()
+	return tw.Close()
+}
+
+func createTarGzipHeaderOnly(output string, hdr *tar.Header) error {
+	out, err := os.Create(output)
+	if err != nil {
+		return err
+	}
+	gw := gzip.NewWriter(out)
+	tw := tar.NewWriter(gw)
+	err = tw.WriteHeader(hdr)
+	if err != nil {
+		return err
+	}
+	err = tw.Close()
+	if err != nil {
+		return err
+	}
+	err = gw.Close()
+	if err != nil {
+		return err
+	}
+	err = out.Close()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func getHeaderFromFileInfo(fi os.FileInfo) (*tar.Header, error) {
+	// TODO: FileInfoHeader gets as much as it can. Change to explicitly
+	// pick only the metadata we care about.
+	return tar.FileInfoHeader(fi, "")
+}
+
+func compressGzip(args fullfileCompressArgs) (int64, error) {
```

Comment:
We can probably abstract this a little more, after we have got things set up we want to

copy the data from an io.Reader to an io.Writer
get the "size" of the io.Writer
close the io.Writer
return the size
Do all of this with proper error handling.

For writing just the header, we can pass in a io.Discard (think /dev/null).

Reply:
I did some factoring of redundant code by not trying to fit both external and internal compressors in the same bucket. I'll handle external compressors with a different function type.

I haven't tried hard to factor out the handling for non-regular files, but will revisit that later.